ENH: Allow users to increase the QHull simplex assignment tolerance

DWesl opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message.
However, please only include an issue number in the description, not the title,
and please ensure that any code names containing underscores are enclosed in backticks.

Depending on your changes, you can skip CI operations and save time and energy: 
https://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->
Closes #21910
Replaces #23285

#### What does this implement/fix?
<!--Please explain your changes.-->
Filling an area from a dense boundary produces holes.  Increasing the tolerance QHull uses to determine whether each test point is in a given simplex allows QHull to assign a simplex to each test point, avoiding those holes.

This PR adds an argument to the `__call__` method of LinearNDInterpolater and CloughTocher2dInterpolator to allow users to increase the tolerance in cases like this.  It also exposes that argument through the `griddata` interface.

<!--
#### Additional information
<!--Any additional information you think is important.-->

DWesl committed:
ENH: Allow users to increase the QHull simplex assignment tolerance.

The default doesn't work for filling a region from dense boundaries.

DOC: Add documentation for the simplex_tolerance parameter.

STY: Wrap long lines

Apply suggestions from code review

Co-authored-by: Evgeni Burovski <evgeny.burovskiy@gmail.com>

DOC: Add versionadded directives to simplex_tolerance and mention
default value.

ev-br reviewed (changes requested):
Looks very reasonable!
Several small requests below: the main one is whether `eps_broad` should be use the new multiplier or its square root.

DWesl commented on scipy/interpolate/src/__fitpack.cc:
```diff
@@ -1,3 +1,5 @@
+#include <Python.h>
+/* __fitpack.h includes Python.h */
```
Would rebase or revert work better for separating out the commits?

ev-br replied:
Either way is likely fine, I'd say.
My personal workflow would be to rebase, but feel free to use what's more convenient to you.

Longer version: the main requirement is that commits that land in main are self-contained. This PR has already 7 commits including a merge from main, so it'll need to be squash-merged. This whatever allows the squash-merge is fine.

DWesl replied:
Commits removed in squash-rebase

DWesl commented:
I meant to ask, do you know why the `_do_evaluate` methods are `def` rather than `cdef`?  `cdef` methods would be hard to override from python, but would be slightly faster.  The dispatching goes through `_evaluate_double` and `_evaluate_complex` anyway, so there's a workaround there, but the speedup would only be noticeable if someone is calling the interpolator in an inner loop somewhere, and the vectorization takes care of the obvious cases already.

ev-br commented on scipy/interpolate/_interpnd.pyx:
```diff
@@ -337,8 +340,8 @@ class LinearNDInterpolator(NDInterpolatorBase):
-        eps = 100 * DBL_EPSILON
-        eps_broad = sqrt(DBL_EPSILON)
+        eps = 100 * DBL_EPSILON * simplex_tolerance
+        eps_broad = sqrt(DBL_EPSILON) * simplex_tolerance
```
Exactly. So IOW the question is whether here we're scaling the factor of 100 in `tol` or `finfo(double).eps`. I have to admit I don't have an intuition for which one it is,  and I'm ready to roll with a conclusion from a domain expert. You probably looked at these things more than anyone else over last several years at least; in your experiments in https://github.com/scipy/scipy/issues/21910, which of the conditions was active, `eps` or `eps_broad`?
for viz, cc [paulbrodersen](https://github.com/paulbrodersen) as the OP and @tylerjereddy as the spatial expert

ev-br commented on scipy/interpolate/tests/test_interpnd.py:
```diff
@@ -216,6 +216,28 @@ def do_interp(interpolator, slice_rows, slice_cols):
+             [0, np.zeros(100), np.arange(400, 500), 800]]
+        )
+        values = np.ones(202)
+        interpolator = interpnd.LinearNDInterpolator(coords.T, values)
```
A very minor ask if you're updating the PR (might do as a follow-up as well): would be nice to `@pytest.parametrize("interpolator_cls", [LinearNDInterpolator, CloughTocher2DInterpolator])` the test, too.

DWesl replied:
Should I try combining the other tests for LinearNDInterpolator and CloughTocher2DInterpolator as well?  There's a few specific to one or the other, but several would seem to apply to both.

ev-br reviewed (commented):
> I meant to ask, do you know why the _do_evaluate methods are def rather than cdef? cdef methods would be hard to override from python, but would be slightly faster. The dispatching goes through _evaluate_double and _evaluate_complex anyway, so there's a workaround there, but the speedup would only be noticeable if someone is calling the interpolator in an inner loop somewhere, and the vectorization takes care of the obvious cases already.

No idea TBH. Might be worth checking if tests rely on them being `def`, and measuring the perf change from making them `cdef` or `cpdef`, yes. In a follow-up PR though.

> Changing the _do_evaluate methods from def to cdef would change the API, so should probably not go into this PR.

Since `_do_evaluate` is private (has a leading underscore), it won't change the official API. There is some chance that somebody is reaching out to them, yes. Then we'll need to weight the performance improvement "for everybody" with how bad the breakage is for this somebody. Thus yes, a separate PR.

------------------------

One other follow-up is a DOC update (https://github.com/scipy/scipy/pull/23285#pullrequestreview-2990743293) to discuss this new `simplex_tolerance` argument. It'd be best to have an illustration, via a worked example where non-default simplex_tolerance helps. One of MREs from gh-23285 or what you discussed in https://github.com/scipy/scipy/pull/23285 can work, I suppose. But that also can be a follow-up PR.

DWesl committed:
TST: Move the boundary test to its own parametrized function

instead of copying it as a method into two classes.

ev-br reviewed (commented):
My comments are all addressed. The only remaining minor point is that if we have two simplex-related tolerances, `eps` and `eps_broad`, where the second one is the square root of the first one; now, if we add a user-adjustable multiplier for the first one, should the second one have the same multiplier or its root, too? Superficially it feels like yes, but I don't have intuition here, and am ready to roll with any suggestion from people more versed in spatial geometry.
Either way, I believe this PR is an improvement over the status quo, as demonstrated by the new test. Thus I think we should merge this PR --- with or without the square root --- for SciPy 1.17.

ev-br committed:
Update scipy/interpolate/_ndgriddata.py

tylerjereddy commented on scipy/interpolate/_interpnd.pyx:
```diff
@@ -151,8 +151,8 @@ class NDInterpolatorBase:
-    def __call__(self, *args):
-        """
+    def __call__(self, *args, simplex_tolerance=1.0):
+        """\
```
is the `\` really needed here?

tylerjereddy commented on scipy/interpolate/_ndgriddata.py:
```diff
@@ -214,6 +214,11 @@ def griddata(points, values, xi, method='linear', fill_value=np.nan,
+        Multiplier for the default tolerance QHull uses to assign
+        a simplex to the xi.  Default is 1.0
+
+        .. versionadded:: 1.17.0
```
same here maybe re: > 0 guard

tylerjereddy commented:
In fact, in my hands on this branch, I can literally purge the entire degenerate simplex block that uses `eps_broad` (diff below the fold) and `spin test -j 12` passes. Yikes.

<details>

```diff
--- a/scipy/spatial/_qhull.pyx
+++ b/scipy/spatial/_qhull.pyx
@@ -1334,40 +1334,6 @@ cdef int _find_simplex_bruteforce(DelaunayInfo_t *d, double *c,
             inside = _barycentric_inside(d.ndim, transform, x, c, eps)
             if inside:
                 return isimplex
-        else:
-            # transform is invalid (nan, implying degenerate simplex)
-
-            # we replace this inside-check by a check of the neighbors
-            # with a larger epsilon
-
-            for k in range(d.ndim+1):
-                ineighbor = d.neighbors[(d.ndim+1)*isimplex + k]
-                if ineighbor == -1:
-                    continue
-
-                transform = d.transform + ineighbor*d.ndim*(d.ndim+1)
-                if transform[0] != transform[0]:
-                    # another bad simplex
-                    continue
-
-                _barycentric_coordinates(d.ndim, transform, x, c)
-
-                # Check that the point lies (almost) inside the
-                # neighbor simplex
-                inside = 1
-                for m in range(d.ndim+1):
-                    if d.neighbors[(d.ndim+1)*ineighbor + m] == isimplex:
-                        # allow extra leeway towards isimplex
-                        if not (-eps_broad <= c[m] <= 1 + eps):
-                            inside = 0
-                            break
-                    else:
-                        # normal check
-                        if not (-eps <= c[m] <= 1 + eps):
-                            inside = 0
-                            break
-                if inside:
-                    return ineighbor
     return -1
 
 cdef int _find_simplex_directed(DelaunayInfo_t *d, double *c,
```

</details>

DWesl committed:
BUG: Raise a ValueError on invalid simplex_tolerance values.

Given the use, it should be maybe < 1 rather than <=0.

DWesl committed:
Merge branch 'griddata-tolerance' of github.com:DWesl/scipy into griddata-tolerance

DWesl committed:
DOC: Describe intended use for simplex_tolerance.

ev-br commented:
Close/reopen + merge main to flush the CI.

ev-br committed:
Merge branch 'main' into griddata-tolerance

ev-br commented:
Looked at this again. I hope it's not too much to ask @DWesl to check if this fixes https://github.com/scipy/scipy/issues/22831 and https://github.com/scipy/scipy/issues/21279, too? 
Both seem to have MWEs. If these can be fixed with a `simplex_tolerance`, it'd be great to have their reproducers added as regression tests.
This will also hopefully help settling the `eps_broad` business. If all known issues can be fixed without touching `eps_broad` at all, then indeed we don't need it.

DWesl committed:
TST: Add the reproducer from #22831 as a regression test

DWesl committed:
BUG: Interpolators aren't imported into main namespace

DWesl committed:
STY: Wrap long lines

DWesl committed:
BUG: Avoid setting multiprocessing default to forkserver on Cygwin.

Setting this is currently failing on Cygwin

DWesl committed:
Merge branch 'griddata-tolerance' of DWesl/scipy into griddata-tolerance

DWesl committed:
TST: Add a test corresponding to issue #21279.

It turns out that the bug in #21279 depends only on the points at indexes
[1, 79, 81, 82].

DWesl committed:
STY: Revert the changes to eps_broad, since they don't affect results

DWesl commented:
I think I have both reproducers in the tests now: #21279 could reproduce with only four points, which simplified things to a point I know how to handle.

(Oddly, I had to use the JSON version of those points: if I just used the NumPy repr of those points, the test passed fine.)

The last two feel like they'd work as hypothesis tests, though I don't know how to constrain the inputs to actually work with a fixed tolerance.

ev-br reviewed (approved):
Tyler's request to remove non-essential changes, https://github.com/scipy/scipy/pull/23737#issuecomment-3509112500, has been implemented: the PR now adds a `simplex_tolerance` multiplier to `eps` and does not touch `eps_broad`; this makes my question about the `eps_broad` multiplier moot. Especially given that indeed that adding `simplex_tolerance` of the order of 3-5 fixes gh-21279 and gh-22831.

The code looks nice sand minimal; the API change is backwards compatible in that it only adds a new keyword argument with the backcompatible default. 

Let's merge this now. Closing three tickets with a single patch feels good.
Thank you @DWesl, thanks @tylerjereddy for the review.

ev-br committed:
Apply suggestions from code review
